lib/src/builder/build_step.dart

```diff
@@ -0,0 +1,80 @@
+// Copyright (c) 2016, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+library build.src.builder.build_step;
+
+import 'dart:async';
+import 'dart:collection';
+import 'dart:convert';
+
+import '../asset/asset.dart';
+import '../asset/id.dart';
+import '../asset/reader.dart';
+import '../asset/writer.dart';
+
+/// A single step in the build processes. This represents a single input and
+/// its expected and real outputs. It also handles tracking of dependencies.
+class BuildStep {
+  /// The primary input for this build step.
+  final Asset input;
+
+  /// The list of all outputs which are expected/allowed to be output from this
+  /// step.
+  final List<AssetId> expectedOutputs;
+
+  /// The actual outputs of this build step.
+  UnmodifiableListView<Asset> get outputs => new UnmodifiableListView(_outputs);
```

Comment:
It feels wrong to me that `outputs` and `dependencies` are "Unmodifiable" but may observably change. Can we hide access to those lists until we are sure they won't change?